File: android/app/src/test/java/cat/plou/alarm/WindowTest.kt
package cat.plou.alarm

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.ZoneId
import java.time.ZonedDateTime

class WindowTest {

    /** Instante en hora local de Madrid, para leer las pruebas con claridad. */
    private fun madrid(y: Int, m: Int, d: Int, h: Int, min: Int): Long =
        ZonedDateTime.of(y, m, d, h, min, 0, 0, ZoneId.of("Europe/Madrid")).toInstant().toEpochMilli()

    private val tz = "Europe/Madrid"

    @Test
    fun `franja normal dentro y fuera`() {
        // Martes 8 de julio de 2025.
        assertTrue(isWithinWindow(madrid(2025, 7, 8, 10, 0), tz, "08:00", "20:00"))
        assertFalse(isWithinWindow(madrid(2025, 7, 8, 21, 0), tz, "08:00", "20:00"))
        assertFalse(isWithinWindow(madrid(2025, 7, 8, 7, 59), tz, "08:00", "20:00"))
    }

    @Test
    fun `el final de la franja es exclusivo`() {
        assertFalse(isWithinWindow(madrid(2025, 7, 8, 20, 0), tz, "08:00", "20:00"))
        assertTrue(isWithinWindow(madrid(2025, 7, 8, 8, 0), tz, "08:00", "20:00"))
    }

    @Test
    fun `franja que cruza la medianoche`() {
        assertTrue(isWithinWindow(madrid(2025, 7, 8, 23, 30), tz, "22:00", "07:00"))
        assertTrue(isWithinWindow(madrid(2025, 7, 9, 3, 0), tz, "22:00", "07:00"))
        assertFalse(isWithinWindow(madrid(2025, 7, 9, 12, 0), tz, "22:00", "07:00"))
    }

    @Test
    fun `la madrugada cuenta como el dia en que empezo la franja`() {
        // Viernes 11 de julio a las 23:00 y madrugada del sábado 12 a las 02:00:
        // ambas pertenecen a la noche del viernes (día 5).
        val viernesNoche = listOf(5)
        assertTrue(isWithinWindow(madrid(2025, 7, 11, 23, 0), tz, "22:00", "07:00", viernesNoche))
        assertTrue(isWithinWindow(madrid(2025, 7, 12, 2, 0), tz, "22:00", "07:00", viernesNoche))
        // La madrugada del viernes pertenece a la noche del jueves, no del viernes.
        assertFalse(isWithinWindow(madrid(2025, 7, 11, 2, 0), tz, "22:00", "07:00", viernesNoche))
    }

    @Test
    fun `sin dias o con los siete no se restringe`() {
        assertTrue(isWithinWindow(madrid(2025, 7, 8, 10, 0), tz, "08:00", "20:00", emptyList()))
        assertTrue(isWithinWindow(madrid(2025, 7, 8, 10, 0), tz, "08:00", "20:00", (0..6).toList()))
    }

    @Test
    fun `una franja vacia no se aplica nunca`() {
        assertFalse(isWithinWindow(madrid(2025, 7, 8, 10, 0), tz, "10:00", "10:00"))
    }

    @Test
    fun `una zona horaria desconocida cae a UTC en vez de fallar`() {
        val at = ZonedDateTime.of(2025, 7, 8, 10, 0, 0, 0, ZoneId.of("UTC")).toInstant().toEpochMilli()
        assertTrue(isWithinWindow(at, "Zona/Inventada", "09:00", "11:00"))
    }

    @Test
    fun `la hora local respeta la zona horaria`() {
        val at = madrid(2025, 7, 8, 10, 30)
        assertEquals(LocalTimeInfo(10 * 60 + 30, 2), localTime(at, tz))
        // El mismo instante son las 08:30 UTC.
        assertEquals(8 * 60 + 30, localTime(at, "UTC").minutes)
    }
}

File: server/test/window.test.ts
import { describe, expect, it } from 'vitest';
import { isWithinWindow, localTime } from '../src/alarm/window.js';

/** Instante UTC concreto: martes 15 de julio de 2025, 21:30 UTC. */
const TUESDAY_2130_UTC = Date.UTC(2025, 6, 15, 21, 30);

describe('localTime', () => {
  it('convierte a la zona horaria pedida', () => {
    // Madrid está en UTC+2 en julio: 23:30 del martes.
    const madrid = localTime(TUESDAY_2130_UTC, 'Europe/Madrid');
    expect(madrid.minutes).toBe(23 * 60 + 30);
    expect(madrid.weekday).toBe(2);

    // Nueva York está en UTC-4: 17:30 del mismo martes.
    const ny = localTime(TUESDAY_2130_UTC, 'America/New_York');
    expect(ny.minutes).toBe(17 * 60 + 30);
    expect(ny.weekday).toBe(2);
  });

  it('cae a UTC ante una zona horaria desconocida', () => {
    const value = localTime(TUESDAY_2130_UTC, 'Marte/Olympus_Mons');
    expect(value.minutes).toBe(21 * 60 + 30);
  });
});

describe('isWithinWindow', () => {
  it('acepta franjas normales dentro del mismo día', () => {
    expect(isWithinWindow(TUESDAY_2130_UTC, 'UTC', '21:00', '22:00')).toBe(true);
    expect(isWithinWindow(TUESDAY_2130_UTC, 'UTC', '08:00', '17:00')).toBe(false);
  });

  it('trata el final de la franja como exclusivo', () => {
    const at0900 = Date.UTC(2025, 6, 15, 9, 0);
    expect(isWithinWindow(at0900, 'UTC', '07:00', '09:00')).toBe(false);
    expect(isWithinWindow(at0900, 'UTC', '09:00', '17:00')).toBe(true);
  });

  it('gestiona franjas que cruzan la medianoche', () => {
    // 22:00–07:00 con la hora local a las 23:30 en Madrid.
    expect(isWithinWindow(TUESDAY_2130_UTC, 'Europe/Madrid', '22:00', '07:00')).toBe(true);
    // A las 03:00 UTC del miércoles sigue dentro.
    const wednesday0300 = Date.UTC(2025, 6, 16, 3, 0);
    expect(isWithinWindow(wednesday0300, 'UTC', '22:00', '07:00')).toBe(true);
    // A las 12:00 está fuera.
    const noon = Date.UTC(2025, 6, 16, 12, 0);
    expect(isWithinWindow(noon, 'UTC', '22:00', '07:00')).toBe(false);
  });

  it('respeta la lista de días en franjas normales', () => {
    // Martes = 2.
    expect(isWithinWindow(TUESDAY_2130_UTC, 'UTC', '21:00', '22:00', [2])).toBe(true);
    expect(isWithinWindow(TUESDAY_2130_UTC, 'UTC', '21:00', '22:00', [3])).toBe(false);
  });

  it('asocia la madrugada al día en que empezó la franja', () => {
    // Miércoles 03:00 pertenece a la noche del martes (día 2).
    const wednesday0300 = Date.UTC(2025, 6, 16, 3, 0);
    expect(isWithinWindow(wednesday0300, 'UTC', '22:00', '07:00', [2])).toBe(true);
    expect(isWithinWindow(wednesday0300, 'UTC', '22:00', '07:00', [3])).toBe(false);
  });

  it('una lista con los siete días equivale a no restringir', () => {
    expect(isWithinWindow(TUESDAY_2130_UTC, 'UTC', '21:00', '22:00', [0, 1, 2, 3, 4, 5, 6])).toBe(true);
  });

  it('una franja vacía no se aplica nunca', () => {
    expect(isWithinWindow(TUESDAY_2130_UTC, 'UTC', '10:00', '10:00')).toBe(false);
  });
});
